### PR TITLE
Rebase EmbeddingOp on the NNVM branch

### PR DESCRIPTION
**Pull request by flyers**
This refactors the Embedding layer to the new operator API based on NNVM according to the guide in #3248.
The current new operator guide is not that easy to follow for beginners. Hope we will have a more detailed document on it in the future.

### CODE REVIEW

File: src/operator/tensor/indexing_op.cc [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  })
+.set_attr<FCompute>("FCompute<cpu>", EmbeddingOpForward<cpu>)
+.set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseIn{"_backward_embedding"})
+.add_argument("data", "Symbol", "Input data to the EmbeddingOp.")
+.add_argument("weight", "Symbol", "Enbedding weight matrix.")
+.add_arguments(EmbeddingParam::__FIELDS__());
+
+NNVM_REGISTER_OP(_backward_embedding)
+.set_num_inputs(3)
+.set_num_outputs(2)
```
Comment (sxjscience) [inline src/operator/tensor/indexing_op.cc]: The _backward_embedding requires only the `out_grad` and `index`. We could reduce the number of input and output nodes here. We are able to manually adjust the gradient node by setting the "FGradient" attribute. May refer to the implementation of "ElemwiseGradUseIn"

**Comment by sxjscience** [comment]
@flyers Try rebase against the latest nnvm branch (Creating a new PR seems to be the simplest way) and we  can merge this in. I'm thinking to add the (take)[https://docs.scipy.org/doc/numpy/reference/generated/numpy.take.html] operator as an alias of embedding. `Embedding(data, index)` is in fact equivalent to `numpy.take(data, index, axis=0)`

File: src/operator/tensor/indexing_op.h [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+
+    if (req[embedding::kWeight] == kWriteTo || req[embedding::kWeight] == kAddTo) {
+      if (req[embedding::kWeight] == kWriteTo) {
+#ifdef __CUDACC__
+        cudaMemsetAsync(grad_in.dptr_, 0, grad_in.MSize() * sizeof(DType),
+                          Stream<gpu>::GetStream(s));
+#else
+        grad_in = scalar<DType>(0.0f);
+#endif
```
Comment (sxjscience) [inline src/operator/tensor/indexing_op.h]: Seems that the nested macro here causes some compilation problem on windows. We could delete the `cudaMemsetAsync` part and retry.

Comment (flyers) [inline src/operator/tensor/indexing_op.h]: Deleting cudaMemsetAsync still works in Ubuntu.

**Pull request closed without merging**